### DISCUSSION TITLE
HAFS-MOM6 IC coupled using CDEPS

### DISCUSSION THREAD

**Discussion by JohnSteffen-NOAA** [hafs-community/HAFS#90] [Q&A] [answered]
Hello all,

I'm working on the standalone HAFS-MOM6 IC coupled configuration which supplies MOM6 SST to FV3, one-way coupled. My working directory on Orion is here,

/work/noaa/hwrf/save/jsteff/hafs_couplehycom_cdeps/

I followed the README instructions in this directory, which I modified from Ling Liu. However, after install/compile of the HAFS system, the generated executable in ./HAFS/sorc/hafs_forecast.fd/tests/fv3_32bit.exe produces an error after 30s of runtime,

"  879: FATAL from PE   879:  atmosphere_dynamics: error in ccpp_physics_init for group fast_physics: Invalid suite FV3_HAFS_v0_gfdlmp_tedmf_nonsst"

Both Ufuk and Bin suggested that there may be an issue with the HAFS model version or build. Any suggestions on how to check this and correct the error?

*I will note, this configuration runs successfully when using a previously compiled fv3_org.exe from Ling. A copy of the original directory is here,

/work/noaa/hwrf/save/jsteff/HAFS_couplling_orig

And successfully generated output files are moved to here,

/work/noaa/hwrf/save/jsteff/HAFS_couplling_orig/cpld_regional_natl_c192_docn_mom6/output_fv3org

Thank you for your advice!

Best,
John

**Comment by evankalina** [comment, marked as answer]
Hi @JohnSteffen-NOAA! Thanks for posting this question. I'm not able to see your directories (permission denied), but this type of error would mean that the FV3_HAFS_v0_gfdlmp_tedmf_nonsst wasn't included in the fv3 executable at compile time. In the HAFS application-level feature/hafs_couplehycom_cdeps branch, the [sorc/build_forecast.sh script](https://github.com/hafs-community/HAFS/blob/feature/hafs_couplehycom_cdeps/sorc/build_forecast.sh) includes this physics suite in the call to compile.sh. Does your build_forecast.sh script look like the one in the branch? Where did you get your code from?

**Reply by JohnSteffen-NOAA** [reply]
Hi @evankalina , thanks for the prompt reply and suggestions.

First, I have lifted the permissions, so you should be able to see the relevant README file and other directories/files as needed. 

Second, I cloned the feature/hafs_couplehycom_cdeps branch last week (08/03) into my working directory using, 

"git clone -b feature/hafs_couplehycom_cdeps --recursive https://github.com/hafs-community/HAFS.git ./HAFS/"

Compiled HAFS without any errors in the build_forecast.log and linked the generated fv3_32bit.exe to the run directory,

"nohup ./build_forecast.sh > build_forecast.log 2>&1 &"

Finally, the build_forecast.sh seems correct and it is a direct match to the one in the branch. It includes the following,

"# For Data Ocean (DOCN):
./compile.sh "$target" "-DAPP=HAFS_DOCN -DCCPP_SUITES=HAFS_v0_gfdlmp_tedmf_nonsst,HAFS_v0_gfdlmp_tedmf,HAFS_v0_gfdlmp_nocpnsst,HAFS_v0_gfdlmp_nonsst,HAFS_v0_gfdlmp_nocp,HAFS_v0_gfdlmp,HAFS_v0_hwrf_thompson,HAFS_v0_hwrf -D32BIT=ON" 32bit YES NO"

The successful build_forecast.log is below. Nothing seems out of the ordinary.
/work/noaa/hwrf/save/jsteff/hafs_couplehycom_cdeps/HAFS/sorc/build_forecast.log

Thank you!

**Reply by evankalina** [reply]
Thanks @JohnSteffen-NOAA! But unfortunately I still can't see your working directory. I think it is because the following directory (and associated subdirectories) under /work/noaa/hwrf/save/ do not have read and execute permission:

drwxr-s---  10 jsteff   hwrf  4096 Aug  9 15:24 jsteff

BTW, if you want new directories and files to be created with global read permission, you can follow the below procedure. This is optional, of course.

-Open ~/.bashrc
-Add a new line that says "umask 022" (without quotes) just before the following lines:
################################################################################
##  All non-interactive shells will exit on the next line.
-Save the file
-Run "source ~/.bashrc" in any open Orion terminal windows where you want this to take effect. Subsequent new login sessions will automatically source the new ~/.bashrc.
-New files and directories created after this change should be readable by everyone.

**Reply by JohnSteffen-NOAA** [reply]
Hi @evankalina , sorry about that! Directory permissions should be fixed now. Thanks for the tidbit on "umask 022" placed in the .bashrc. Very convenient!

To note, I have moved the err & out files from the failed fv3 atm run to,
/work/noaa/hwrf/save/jsteff/hafs_couplehycom_cdeps/cpld_regional_natl_c192_docn_mom6/output_fv3_32bit_err

**Reply by evankalina** [reply]
@JohnSteffen-NOAA Hmm, it still looks unreadable to me:

ls -ltr /work/noaa/hwrf/save/
drwxr-s---  10 jsteff   hwrf  4096 Aug 13 13:03 jsteff

The following command should open global read+execute permission on the jsteff directory (along with any subdirectories):
cd /work/noaa/hwrf/save/
find jsteff -type d -exec chmod 755 {} \;

Unfortunately it does take some time to finish running. After it finishes, you may want to check that the files also have global read permission.

**Reply by JohnSteffen-NOAA** [reply]
@evankalina Okay, corrected. Dir and files have 755 permission throughout.

**Reply by evankalina** [reply]
@JohnSteffen-NOAA I can see your directory now – thanks! OK, in your input.nml file, you should set ccpp_suite = 'HAFS_v0_gfdlmp_tedmf_nonsst'. It currently has "FV3_" in front of it. I believe this works with Ling's executable because she is using a different version of the source code, but the feature/hafs_couplehycom_cdeps branch defines the various suites without the leading "FV3_" (see the build_forecast.sh script). I need to update this feature branch with the latest commits from develop, and then everything should be standardized. But for now, this fix should work for you.

**Reply by JohnSteffen-NOAA** [reply]
Thanks @evankalina  , I'll give it a try.

**Reply by JohnSteffen-NOAA** [reply]
Good morning @evankalina , I was able to test your fix this weekend. It did resolve the issue with the CCPP physics suites, however, I am returning a new error related to the PIO library. Fv3 runs for about 2.5 minutes before this err from the logfile,

1282: Abort with message No such file or directory in file /apps/contrib/NCEP/src/hpc-stack-v1.1.0/pkg/pio-2.5.2/src/clib/pioc_support.c at line 2747

I scrolled through the pioc_support.c file and this is the code around line 2747 that calls check_netcdf2,

    /* If there was an error, free allocated memory and deal with the error. */
    if (ierr)
    {
        free(file);
        **return check_netcdf2(ios, NULL, ierr, __FILE__, __LINE__);**
    }

Any thoughts on the potential cause of this? Thanks.

**Reply by evankalina** [reply]
@JohnSteffen-NOAA Thanks for letting me know. Here are two thoughts:

1) In the model_configure file, would you try repeating the run with print_esmf: .true.? This will generate a lot of diagnostic output, which should help us understand whether this is really a problem with the PIO library or if something more basic is wrong.
2) As of today, I updated the feature/hafs_couplehycom_cdeps branch with the latest code from the develop branch at the application level and from the support/HAFS branches of the various submodules. I believe the resulting code is closer to the one that Ling Liu has been using, although it is probably more up to date than hers as well. You may want to try checking out the latest version of this branch and seeing if it impacts this problem. If you choose to do this, you will want to update your input.nml to add the "FV3_" back to the ccpp_suite, since that is what the suite is called in the latest version of ufs-weather-model.

**Reply by JohnSteffen-NOAA** [reply]
Hi @evankalina ,

I am provide an update on the DOCN - MOM6 IC run. I attempted to run this configuration following your directions above and I am getting a new error, shown below. I updated the model_configure file to print_esmf: .true and this outputs the PET####logfile which I have collected in /cpld_regional_natl_c192_docn_mom6/PETlog/ . Also, I did a "git pull" on the feature/hafs_couplehycom_cdeps/ to update my HAFS directory. I recompiled the code successfully and linked the fv3_32bit.exe file before running. The model fails after ~30s runtime. I also attempted to run this with and without the "FV3_" and I get the same error. Thanks for your help!


1112: FATAL from PE  1112: check_nml_error in fms_mod:  Unknown namelist, or mistyped namelist variable in namelist atmos_model_nml, (IOSTAT =           19 )

**Reply by evankalina** [reply]
Hi @JohnSteffen-NOAA,

I believe this error is being triggered by some now defunct settings in your input.nml file. Specifically, the latest version of ufs-weather-model no longer uses these input.nml variables in &atmos_model_nml: fdiag, fhmax, fhout, fhmaxhf, fhouthf. Also, merge_import has been renamed to cpl_imp_mrg and moved to the &gfs_physics_nml. And yes, you should include the FV3_ in the ccpp_suite with this new version of the code.

Please see these commits for more information:
https://github.com/ufs-community/ufs-weather-model/commit/74a34d225fb76702e48c0a27534514992ac2ce77
https://github.com/NOAA-EMC/fv3atm/commit/6ac48b1d36b5559500dcb4718bae9140206fc2b8

**Reply by JohnSteffen-NOAA** [reply]
Thanks @evankalina , I will give this a try and refer to the commits you linked. Have a nice weekend!

**Reply by JohnSteffen-NOAA** [reply]
Good morning @evankalina , now that the settings are correct in the input.nml file, the error related to the unknown namelist variable in atmos_model_nml is resolved. Thanks! 

 However, I am still receiving the original error related to the PIO library from last week. Here is the output from err file again. The print_esmf: .true. is set and additional information is in the PET####.ESMF_LogFile located in dir, /work/noaa/hwrf/save/jsteff/hafs_couplehycom_cdeps/cpld_regional_natl_c192_docn_mom6/PETlog


1282: Abort with message No such file or directory in file /apps/contrib/NCEP/src/hpc-stack-v1.1.0/pkg/pio-2.5.2/src/clib/pioc_support.c at line 2747

Could this be related to my module environment? I'm using the job_card provided by Ling to submit the job, which uses the following module setup. 

source ./module-setup.sh
module use $( pwd -P )
module load modules.fv3
module list

Currently Loaded Modules:
  1) contrib/0.1       13) png/1.6.35      25) w3nco/2.4.1
  2) noaatools/2.0     14) hdf5/1.10.6     26) nemsio/2.5.2
  3) cmake/3.18.1      15) netcdf/4.7.4    27) w3emc/2.7.3
  4) python/3.7.5      16) pio/2.5.2       28) crtm/2.3.0
  5) hpc/1.1.0         17) esmf/8_1_1      29) upp/10.0.8
  6) intel/2018.4      18) fms/2020.04.03  30) gftl-shared/v1.3.0
  7) mkl/2020.2        19) g2/3.4.3        31) yafyaml/v0.5.1
  8) hpc-intel/2018.4  20) g2tmpl/1.10.0   32) mapl/v2.7.3
  9) impi/2018.4       21) ip/3.3.3        33) ufs_common
 10) hpc-impi/2018.4   22) sp/2.3.3        34) modules.fv3
 11) jasper/2.0.22     23) sigio/2.3.2
 12) zlib/1.2.11       24) bacio/2.4.1

**Reply by lliu1220** [reply]
@JohnSteffen-NOAA 
Try the following to build:

git clone -b feature/hafs_couplehycom_cdeps --recursive https://github.com/hafs-community/HAFS
git clone -b develop --recursive https://github.com/hafs-community/HAFS
cd HAFS/sorc/hafs_forecast.fd
git checkout feature/hafs_couplehycom_cdeps
git submodule update --init —recursive

and then you could compile the model by amodifying build_forecast.sh as following,

./compile.sh "$target" "SUITES=HAFS_v0_gfdlmp_nonsst 32BIT=Y CDEPS=Y CDEPS_DOCN=Y CMEPS=Y" 32bit YES NO

then you could link those files to your run directory,

cd RUN_test_fv3_docn
ln -sf $YOUR_PATH/HAFS/sorc/hafs_forecast.fd/NEMS/src/conf/module-setup.sh.inc module-setup.sh
ln -sf $YOUR_PATH/HAFS/sorc/hafs_forecast.fd/modulefiles/orion.intel/fv3 modules.fcst
ln -sf $YOUR_PATH/HAFS/sorc/hafs_forecast.fd/tests/fv3_32bit.exe .

**Reply by JohnSteffen-NOAA** [reply]
Hi @lliu1220 , with the recent updates made to feature/hafs_couplehycom_cdeps some of these steps are outdated and no longer work. Maybe @uturuncoglu could provide some insight. Thank you all for the help!

**Reply by evankalina** [reply]
@JohnSteffen-NOAA It appears that your input data file doesn't exist:

Orion-login-1[39] ekalina$ pwd
/work/noaa/hwrf/save/jsteff/hafs_couplehycom_cdeps/cpld_regional_natl_c192_docn_mom6/INPUT_DOCN
Orion-login-1[40] ekalina$ ls -ltr hycom_201908_surf_nolev.nc
lrwxrwxrwx 1 jsteff hwrf 80 Aug  2 13:07 hycom_201908_surf_nolev.nc -> ../../old_RUN_test_fv3_docn/DOCN_MOM6_input_data/hycom_201908_surf_nolev_mom6.nc
Orion-login-1[41] ekalina$ ls -ltr ../../old_RUN_test_fv3_docn/DOCN_MOM6_input_data/hycom_201908_surf_nolev_mom6.nc
ls: cannot access ../../old_RUN_test_fv3_docn/DOCN_MOM6_input_data/hycom_201908_surf_nolev_mom6.nc: No such file or directory

**Reply by JohnSteffen-NOAA** [reply]
Hi @evankalina , that is an oversight by me. It's a broken link from when I initially copied Ling's directory. I have it set up now where the input files from HYCOM and MOM6 are located in my working directory and linked from there. I'll add this step to the README file and this should provide all the necessary steps to test this configuration. 

I ran the fv3_32bit.exe and it works. There is a NetCDF start+count error towards the end of the run, but that is a separate issue I have already discussed with Ling.

Thanks for all your help! I appreciate it.

Best,
John

**Reply by evankalina** [reply]
@JohnSteffen-NOAA Great to hear that! Please let me know if you run into any other issues.

**Comment by YongzuoLi-NOAA** [comment]
Hi @JohnSteffen-NOAA I am reading this. My working directory on Orion is here, /work/noaa/hwrf/save/jsteff/hafs_couplehycom_cdeps/ I followed the README instructions in this directory, which I modified from Ling Liu.
But I can not find README. README.md there has no Ling Liu modification.

**Reply by JohnSteffen-NOAA** [reply]
Hi @YongzuoLi , try this directory here, /work/noaa/hwrf/save/jsteff/hafs_couplehycom_cdeps_mom6 . This work was performed before a HAFS-CDEPS approved PR, so further modifications may be needed. There's an updated HAFS-CDEPS configuration which can import/use DOCN in hycom formatting. What are you trying to do?

**Comment by YongzuoLi-NOAA** [comment]
Hi @JohnSteffen-NOAA Thank you for the information. I now see /work/noaa/hwrf/save/jsteff/hafs_couplehycom_cdeps_mom6/README and will follow it. I now work on regional MOM6-CDEPS-HAFS. Are your HAFS and MOM6 (SIS2 GFDL repo ?) here for regional ?

**Reply by JohnSteffen-NOAA** [reply]
@YongzuoLi Yes, these configurations are setup for regional HAT10 domain. I would suggest following up with Cameron Book. He is also currently working on HAFS-CDEPS runs using MOM6 IC.

**Reply by YongzuoLi-NOAA** [reply]
@JohnSteffen-NOAA Yes, I am working Cameron Book now.
